File: rust/logbrew/src/operation_tracing.rs
use crate::{
    Metadata, OpenTelemetrySpanContext, SdkError, SpanEvent, Traceparent, TraceparentContext,
};
use serde_json::{Map, Value};

const UNSAFE_KEY_PARTS: &[&str] = &[
    "authorization",
    "body",
    "broker",
    "command",
    "connection",
    "cookie",
    "dsn",
    "header",
    "host",
    "jobid",
    "key",
    "message",
    "param",
    concat!("pass", "word"),
    "payload",
    "query",
    concat!("sec", "ret"),
    "sql",
    "statement",
    concat!("to", "ken"),
    "url",
    "user",
    "value",
];

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
/// Dependency operation category used for low-cardinality span metadata.
pub enum DependencyOperationKind {
    /// Database, ORM, or query-like work owned by the app.
    Database,
    /// Cache, key-value, or memoization work owned by the app.
    Cache,
    /// Queue, job, stream, or broker work owned by the app.
    Queue,
}

impl DependencyOperationKind {
    fn source(self) -> &'static str {
        match self {
            Self::Database => "database.operation",
            Self::Cache => "cache.operation",
            Self::Queue => "queue.operation",
        }
    }

    fn prefix(self) -> &'static str {
        match self {
            Self::Database => "db",
            Self::Cache => "cache",
            Self::Queue => "messaging",
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
/// Explicit dependency span builder for DB, cache, and queue work.
pub struct DependencyOperationSpan {
    kind: DependencyOperationKind,
    name: String,
    span_id: String,
    status: String,
    duration_ms: Option<f64>,
    system: Option<String>,
    operation: Option<String>,
    target: Option<String>,
    error_type: Option<String>,
    metadata: Metadata,
}

impl DependencyOperationSpan {
    /// Create a database operation span builder with app-owned low-cardinality IDs.
    pub fn database(name: impl Into<String>, span_id: impl Into<String>) -> Self {
        Self::new(DependencyOperationKind::Database, name, span_id)
    }

    /// Create a cache operation span builder with app-owned low-cardinality IDs.
    pub fn cache(name: impl Into<String>, span_id: impl Into<String>) -> Self {
        Self::new(DependencyOperationKind::Cache, name, span_id)
    }

    /// Create a queue operation span builder with app-owned low-cardinality IDs.
    pub fn queue(name: impl Into<String>, span_id: impl Into<String>) -> Self {
        Self::new(DependencyOperationKind::Queue, name, span_id)
    }

    fn new(
        kind: DependencyOperationKind,
        name: impl Into<String>,
        span_id: impl Into<String>,
    ) -> Self {
        Self {
            kind,
            name: name.into(),
            span_id: span_id.into(),
            status: "ok".to_string(),
            duration_ms: None,
            system: None,
            operation: None,
            target: None,
            error_type: None,
            metadata: Map::new(),
        }
    }

    /// Override the operation status. Valid values are `ok` and `error`.
    pub fn with_status(mut self, status: impl Into<String>) -> Self {
        self.status = status.into();
        self
    }

    /// Attach an optional non-negative duration in milliseconds.
    pub fn with_duration_ms(mut self, duration_ms: f64) -> Self {
        self.duration_ms = Some(duration_ms);
        self
    }

    /// Attach the dependency system, such as `postgres`, `redis`, or `sqs`.
    pub fn with_system(mut self, system: impl Into<String>) -> Self {
        self.system = Some(system.into());
        self
    }

    /// Attach the low-cardinality operation, such as `select`, `get`, or `publish`.
    pub fn with_operation(mut self, operation: impl Into<String>) -> Self {
        self.operation = Some(operation.into());
        self
    }

    /// Attach a low-cardinality target, such as a table, cache region, or queue name.
    pub fn with_target(mut self, target: impl Into<String>) -> Self {
        self.target = Some(target.into());
        self
    }

    /// Mark the span as failed and attach only the public exception type name.
    pub fn with_error_type(mut self, error_type: impl Into<String>) -> Self {
        self.status = "error".to_string();
        self.error_type = Some(error_type.into());
        self
    }

    /// Attach primitive app-owned metadata; unsafe key names and non-primitives are dropped.
    pub fn with_metadata(mut self, metadata: Metadata) -> Self {
        self.metadata = sanitized_metadata(metadata);
        self
    }

    /// Build a normal LogBrew span event from an incoming W3C traceparent context.
    pub fn from_traceparent_context(
        self,
        context: &TraceparentContext,
    ) -> Result<SpanEvent, SdkError> {
        let metadata = self.metadata();
        let mut span = Traceparent::span_attributes_from_context(
            context,
            crate::TraceparentSpanInput::new(self.span_name(), self.span_id, self.status),
        )?;
        if let Some(duration_ms) = self.duration_ms {
            span = span.with_duration_ms(duration_ms);
        }
        Ok(span.with_metadata(metadata))
    }

    /// Build a normal LogBrew span event from an OpenTelemetry-compatible current span.
    pub fn from_opentelemetry_context(
        self,
        context: &OpenTelemetrySpanContext,
    ) -> Result<SpanEvent, SdkError> {
        let trace_context = Traceparent::context_from_opentelemetry_context(context);
        self.from_traceparent_context(&trace_context)
    }

    fn span_name(&self) -> String {
        format!("{}:{}", self.kind.source(), self.name.trim())
    }

    fn metadata(&self) -> Metadata {
        let mut metadata = self.metadata.clone();
        metadata.insert(
            "source".to_string(),
            Value::String(self.kind.source().to_string()),
        );
        insert_string(
            &mut metadata,
            format!("{}.system", self.kind.prefix()),
            &self.system,
        );
        insert_string(
            &mut metadata,
            format!("{}.operation", self.kind.prefix()),
            &self.operation,
        );
        insert_string(
            &mut metadata,
            format!("{}.target", self.kind.prefix()),
            &self.target,
        );
        insert_string(&mut metadata, "exception.type", &self.error_type);
        metadata
    }
}

fn insert_string(metadata: &mut Metadata, key: impl Into<String>, value: &Option<String>) {
    if let Some(value) = value
        .as_deref()
        .map(str::trim)
        .filter(|value| !value.is_empty())
    {
        metadata.insert(key.into(), Value::String(value.to_string()));
    }
}

fn sanitized_metadata(metadata: Metadata) -> Metadata {
    metadata
        .into_iter()
        .filter(|(key, value)| is_safe_key(key) && value_is_primitive(value))
        .collect()
}

fn is_safe_key(key: &str) -> bool {
    let normalized = normalized_key(key);
    !UNSAFE_KEY_PARTS
        .iter()
        .any(|part| normalized.contains(part))
}

fn normalized_key(key: &str) -> String {
    key.chars()
        .filter(|character| character.is_ascii_alphanumeric())
        .flat_map(char::to_lowercase)
        .collect()
}

fn value_is_primitive(value: &Value) -> bool {
    matches!(
        value,
        Value::Null | Value::Bool(_) | Value::Number(_) | Value::String(_)
    )
}

File: ruby/logbrew-ruby/lib/logbrew/operation_tracing.rb
# frozen_string_literal: true

module LogBrew
  # Explicit dependency spans for app-owned database, cache, and queue work.
  module OperationTracing
    UNSAFE_METADATA = /authorization|body|broker|cache.?key|command|connection|cookie|dsn|header|host|
      \bjid\z|job.?id|key|message|param|pass#{'word'}|payload|query|sec#{'ret'}|sql|statement|
      to#{'ken'}|url|username|value/ix
    private_constant :UNSAFE_METADATA

    module_function

    def database_operation(client, name, **options, &block)
      capture_operation(client, "database", name, options, &block)
    end

    def cache_operation(client, name, **options, &block)
      capture_operation(client, "cache", name, options, &block)
    end

    def queue_operation(client, name, **options, &block)
      capture_operation(client, "queue", name, options, &block)
    end

    def capture_operation(client, kind, name, options)
      raise SdkError.new("validation_error", "#{kind} operation block is required") unless block_given?

      Validation.require_non_empty("#{kind} operation name", name)
      started_at = monotonic_time
      context = child_context
      error = nil
      result = nil

      Trace.with_context(context) do
        begin
          result = yield context
        rescue StandardError => captured
          error = captured
        end
      end

      capture_span(client, kind, name, context, started_at, options, error)
      raise error if error

      result
    end

    def capture_span(client, kind, name, context, started_at, options, error)
      Trace.with_context(context) do
        client.span(
          event_id(kind, context, options),
          timestamp(options),
          {
            name: "#{kind}.operation:#{name}",
            traceId: context.trace_id,
            spanId: context.span_id,
            parentSpanId: context.parent_span_id,
            status: error ? "error" : "ok",
            durationMs: duration_ms(started_at, options),
            metadata: span_metadata(kind, options, error),
            events: span_events(error)
          }
        )
      end
    rescue StandardError => capture_error
      on_error = read_option(options, :on_error)
      begin
        on_error.call(capture_error) if on_error.respond_to?(:call)
      rescue StandardError
        nil
      end
    end

    def child_context
      parent = Trace.current
      return Trace.create_root unless parent

      Trace.create(
        trace_id: parent.trace_id,
        span_id: Trace.generate_span_id,
        parent_span_id: parent.span_id,
        trace_flags: parent.trace_flags
      )
    end

    def span_metadata(kind, options, error)
      sanitized_metadata(read_option(options, :metadata)).tap do |metadata|
        metadata["source"] = read_option(options, :source) || "#{kind}.operation"
        add_option(metadata, "#{kind}.system", read_option(options, :system))
        add_option(metadata, "#{kind}.operation", read_option(options, :operation))
        add_option(metadata, "#{kind}.target", read_option(options, :target))
        metadata["exceptionType"] = error.class.name if error
      end
    end

    def span_events(error)
      return nil unless error

      [
        {
          name: "exception",
          metadata: {
            exceptionType: error.class.name,
            exceptionEscaped: true
          }
        }
      ]
    end

    def sanitized_metadata(metadata)
      return {} if metadata.nil?
      raise SdkError.new("validation_error", "operation metadata must be an object") unless metadata.is_a?(Hash)

      metadata.each_with_object({}) do |(key, value), copied|
        normalized_key = key.to_s
        next if normalized_key.strip.empty?
        next if unsafe_metadata_key?(normalized_key)
        next unless CaptureHelpers.primitive_metadata_value?(value)

        copied[normalized_key] = value
      end
    end

    def unsafe_metadata_key?(key)
      key.match?(UNSAFE_METADATA)
    end

    def add_option(metadata, key, value)
      metadata[key] = value if CaptureHelpers.primitive_metadata_value?(value) && !(value.is_a?(String) && value.strip.empty?)
    end

    def read_option(options, key)
      options.key?(key) ? options[key] : options[key.to_s]
    end

    def event_id(kind, context, options)
      read_option(options, :event_id) || "ruby_#{kind}_span_#{context.span_id}"
    end

    def timestamp(options)
      value = read_option(options, :timestamp)
      return value unless value.nil?

      Time.now.utc.iso8601
    end

    def duration_ms(started_at, options)
      configured = read_option(options, :duration_ms)
      return configured unless configured.nil?

      ((monotonic_time - started_at) * 1000.0).round(3)
    end

    def monotonic_time
      Process.clock_gettime(Process::CLOCK_MONOTONIC)
    end

  end
end
